server/src/main/java/com/linecorp/centraldogma/server/internal/storage/PurgeSchedulingService.java

```diff
@@ -0,0 +1,214 @@
+        }
+    }
+
+    private void cleanPurgedFiles() {
+        projectManager.purgeMarked();
+        projectManager.list().forEach((unused, project) -> project.repos().purgeMarked());
+    }
+
+    @VisibleForTesting
+    void schedulePurgeData(CommandExecutor commandExecutor,
+                           MetadataService metadataService) {
+        final long minAllowedTimestamp = System.currentTimeMillis() - maxRemovedRepositoryAgeMillis;
+        final Predicate<Instant> olderThanMinAllowed =
+                removedAt -> removedAt.toEpochMilli() < minAllowedTimestamp;
+
+        purgeProject(commandExecutor, olderThanMinAllowed);
+        purgeRepository(commandExecutor, metadataService, olderThanMinAllowed);
+    }
+
+    private void purgeProject(CommandExecutor commandExecutor,
+                              Predicate<Instant> olderThanMinAllowed) {
+        projectManager
+                .listRemoved()
+                .forEach((projectName, removal) -> {
+                    if (olderThanMinAllowed.test(removal)) {
+                        commandExecutor
+                                .execute(Command.purgeProject(Author.SYSTEM, projectName)).join();
+                    }
+                });
+    }
+
+    private void purgeRepository(CommandExecutor commandExecutor,
+                                 MetadataService metadataService,
+                                 Predicate<Instant> olderThanMinAllowed) {
+        projectManager
+                .list()
+                .forEach((unused, project) -> {
+                    project.repos().listRemoved()
+                           .forEach((repoName, removal) -> {
+                               if (olderThanMinAllowed.test(removal)) {
+                                   commandExecutor.execute(Command.purgeRepository(Author.SYSTEM,
+                                                                                   project.name(),
+                                                                                   repoName))
+                                                  .join();
+                                   metadataService.purgeRepo(Author.SYSTEM, project.name(), repoName).join();
+                               }
+                           });
+                });
+    }
+
+    private boolean isDisabled() {
+        return maxRemovedRepositoryAgeMillis == 0;
+    }
+
+    private final class StoragePurgingScheduler {
+
+        @Nullable
+        private volatile ListeningScheduledExecutorService scheduler;
+
+        public boolean isStarted() {
+            return scheduler != null;
+        }
+
+        public synchronized void start(Runnable task) {
+            if (isStarted()) {
+                return;
+            }
+            requireNonNull(task, "task");
+            scheduler = MoreExecutors.listeningDecorator(purgeWorker);
+            final ListenableScheduledFuture<?> future = scheduler.scheduleWithFixedDelay(
```

Comment:
How about:
```
final ListeningScheduledExecutorService scheduler = MoreExecutors.listeningDecorator(purgeWorker);
this.scheduler = scheduler;
scheduler.scheduleWithFixedDelay( ...)
```
not to access the volatile field twice?

Reply:
It is a good idea.